avro-builder/builder-spi/src/main/java/com/linkedin/avroutil1/builder/util/StreamUtil.java

```diff
@@ -72,36 +78,27 @@ private StreamUtil() {
    */
   public static <T, R> Collector<T, ?, Stream<R>> toParallelStream(Function<T, R> mapper, int parallelism,
       int batchSize) {
-    if (parallelism <= 0 || batchSize <= 0) {
-      throw new IllegalArgumentException("Parallelism and batch size must be >= 1");
-    }
-
-    return Collectors.collectingAndThen(Collectors.toList(), list -> {
-      if (list.isEmpty()) {
-        return Stream.empty();
-      }
-
-      if (parallelism == 1 || list.size() <= batchSize) {
-        return list.stream().map(mapper);
-      }
+    return new ParallelStreamCollector<>(mapper, parallelism, batchSize);
+  }
 
-      final Executor limitingExecutor = new LimitingExecutor(parallelism);
-      final int batchCount = (list.size() - 1) / batchSize;
-      return IntStream.rangeClosed(0, batchCount)
-          .mapToObj(batch -> {
-            int startIndex = batch * batchSize;
-            int endIndex = (batch == batchCount) ? list.size() : (batch + 1) * batchSize;
-            return list.subList(startIndex, endIndex);
-          })
-          .map(batch -> CompletableFuture.supplyAsync(() -> batch.stream().map(mapper).collect(Collectors.toList()),
-              limitingExecutor))
-          .map(CompletableFuture::join)
-          .flatMap(Collection::stream);
-    });
+  /**
+   * A convenience {@link Collector} used for executing parallel computations on the user provided {@link Executor}
+   * and returning a {@link Stream} instance returning results as they arrive.
+   *
+   * @param mapper a transformation to be performed in parallel
+   * @param batchSize the size into which inputs should be batched before running the mapper.
+   * @param executor executor for executing computations in parallel
+   * @param <T> the type of the collected elements
+   * @param <R> the result returned by {@code mapper}
+   *
+   * @return a {@code Collector} which collects all processed elements into a {@code Stream} in parallel.
+   */
+  public static <T, R> Collector<T, ?, Stream<R>> toParallelStream(Function<T, R> mapper, int batchSize,
+      Executor executor) {
```

Comment:
Not a fan of exposing this. I think we should provide a sensible default in the util. The reason is that if someone uses an improperly sized pool, it can result in a deadlock.

Reply:
Optimized the LimitingExecutor and remove this interface for allowing user pass in executor